default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+    useful_part_classes = ALL_META_CLASSES  # only these parts are considered in the design process
+
+    filter_useful_parts = True              # removes any part not belonging to self.useful_part_classes
+    filter_inefficient_parts = False        # removes cost-inefficient parts (less capacity and less capacity/cost)
+
+    running_index = 1                       # running index used in the design name
+    running_index_needs_update = True       # To check if we need to update the running_index after a load.
+
+    def __init__(self):
+        """Make sure to call this constructor in each subclass."""
+        self.species = None         # name of the piloting species (string)
+        self.hull = None            # hull object (not hullname!)
+        self.partnames = []         # list of partnames (string)
+        self.parts = []             # list of actual part objects
+        self.attacks = {}           # {damage:count}
+        self.structure = 0
+        self.shields = 0
+        self.fuel = 0
+        self.speed = 0
+        self.stealth = 0
+        self.detection = 0
+        self.troops = 0
+        self.colonisation = -1      # -1 since 0 indicates an outpost (capacity = 0)
+        self.production_cost = 9999
+        self.production_time = 1
+        self.pid = -1               # planetID for checks on production cost if not LocationInvariant.
+        self.additional_specifications = AdditionalSpecifications()
+
+    def evaluate(self):
+        """ Return a rating for the design.
+
+        First, check if the minimum requirements are met. If so, return the _rating_function().
+        Otherwise, return a large negative number scaling with distance to the requirements.
+
+        :returns: float - rating of the current part/hull combo
+        """
+        self.update_stats()
+        ret_val = 0
+
+        # If we do not meet the requirements, we want to return a negative rating.
+        # However, we also need to make sure, that the closer we are to requirements,
+        # the better our rating is so the optimizing heuristic finds "the right way".
+        if self.fuel < self.additional_specifications.minimum_fuel:
+            ret_val += -1000 * (self.additional_specifications.minimum_fuel - self.fuel)
+        if self.speed < self.additional_specifications.minimum_speed:
+            ret_val += -1000 * (self.additional_specifications.minimum_speed - self.speed)
+        if self.structure < self.additional_specifications.minimum_structure:
+            ret_val += -1000 * (self.additional_specifications.minimum_structure - self.structure)
+        if ret_val < 0:
+            return ret_val
+        else:
+            return self._rating_function()
+
+    def _rating_function(self):
+        """Rate the design according to current hull/part combo.
+
+        :returns: float - rating
+        """
+        print "WARNING: Rating function not overloaded for class %s!" % self.__class__.__name__
```

Comment:
IMHO game should be stopped on this moment. Since not implementing interface is serious issue and cant be missed on testing.  I suggest to `exit()`. Dilvish, does not agree with such approach, so we just make pokerface and continue gameplay.  But I insist on sending it to human:

```python
from freeorion_tools import print_error
print_error("WARNING: Rating function not overloaded for class %s!" % self.__class__.__name__)
```